gordo/machine/dataset/filter_periods.py

```diff
@@ -0,0 +1,205 @@
+from pprint import pformat
+import pandas as pd
+from sklearn.ensemble import IsolationForest
+from sklearn.preprocessing import MinMaxScaler
+from gordo.machine.dataset.datasets import pandas_filter_rows
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class WrongFilterMethodType(TypeError):
+    pass
+
+
+class filter_periods:
+    """Model class with methods for data pre-processing.
+    Performs a series of algorithms that drops noisy data.
+
+    Either a rolling median or an isolation forest algorithm is executed.
+    Both provide drop periods in a dict-type element on the class object
+    `object.drop_periods["iforest"]` and `object.drop_periods["median"]`,
+    and data is filtered accordingly.
+
+    Parameters:
+    data: pandas.DataFrame
+        Data frame containing already filtered data (global max/min + dropped known periods).
+        Time consecutively is not required.
+    granularity: str
+        The bucket size for grouping all incoming time data (e.g. "10T").
+        Available strings come from https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#dateoffset-objects
+    **kwargs:
+        See below.
+
+    Keyword arguments:
+    filter_method: str
+        Which method should be used for data cleaning, either "median" (default), "iforest" or "all" which returns
+        results for both methods.
+    iforest_smooth: bool
+        If exponential weighted smoothing should be applied to data before isolation forest
+        algorithm is run.
+    """
+
+    def __init__(self, data, granularity, **kwargs):
+        self.data = data.copy()
+        self.granularity = granularity
+        self.filter_method = kwargs.get("filter_method", "median")
+
+        if self.filter_method not in ["median", "iforest", "all"]:
+            raise WrongFilterMethodType
+
+        self.predictions = {}
+        if self.filter_method in ["median", "all"]:
+            self._window = kwargs.get("window", 144)
+            self._n_iqr = kwargs.get("n_iqr", 5)
+            self._rolling_median()
+
+        if self.filter_method in ["iforest", "all"]:
+            self._iforest_smooth = kwargs.get("iforest_smooth", False)
+            self._contamination = kwargs.get("contamination", 0.03)
+            self._train()
+            self._predict()
+
+        self._drop_periods()
+        self._filter_data()
```

Comment:
The class with such behaver looks more like a function then the class. My suggestion is to move all heavy train/predict  logic outside the constructor into separate method `filter_df`
```suggestion
    def __init__(self, granularity, filter_method, window=144, n_iqr=5, iforest_smooth=False, contamination=0.03):
        self.granularity = granularity
        self.filter_method = filter_method

        if self.filter_method not in ["median", "iforest", "all"]:
            raise WrongFilterMethodType

        self._window = window
        self._n_iqr = n_iqr
        self._iforest_smooth = iforest_smooth
        self._contamination =contamination
        
    def filter_df(self, data):
        predictions = {}
        if self.filter_method in ["median", "all"]:
             predictions["median"] = self._rolling_median(data)
             
        if self.filter_method in ["iforest", "all"]:
             self._train(data)
             predictions["iforest"] = self._predict(data)

        drop_periods = self._drop_periods(predictions)
        data = self._filter_data(data, drop_periods)
        return data, drop_periods
```